core/vm/jit.go

```diff
@@ -0,0 +1,592 @@
+		if !context.UseGas(cost) {
+			return nil, OutOfGasError
+		}
+		// Resize the memory calculated previously
+		mem.Resize(newMemSize.Uint64())
+
+		// These opcodes return an argument and are thefor handled
+		// differently from the rest of the opcodes
+		switch instr.op {
+		case JUMP:
+			if err := jump(stack.pop()); err != nil {
+				return nil, err
+			}
+			continue
+		case JUMPI:
+			pos, cond := stack.pop(), stack.pop()
+
+			if cond.Cmp(common.BigTrue) >= 0 {
+				if err := jump(pos); err != nil {
+					return nil, err
+				}
+				continue
+			}
+		case RETURN:
+			offset, size := stack.pop(), stack.pop()
+			ret := mem.GetPtr(offset.Int64(), size.Int64())
+
+			return context.Return(ret), nil
+		case SUICIDE:
+			instr.fn(instr, env, context, mem, stack)
+
+			return context.Return(nil), nil
+		case STOP:
+			return context.Return(nil), nil
+		default:
+			if instr.fn == nil {
+				return nil, fmt.Errorf("Invalid opcode %x", instr.op)
+			}
+
+			instr.fn(instr, env, context, mem, stack)
+		}
+
+		pc++
+	}
+
+	return context.Return(nil), nil
+}
+
+// validDest checks if the given distination is a valid one given the
+// destination table of the program
+func validDest(dests map[uint64]struct{}, dest *big.Int) bool {
+	// PC cannot go beyond len(code) and certainly can't be bigger than 64bits.
+	// Don't bother checking for JUMPDEST in that case.
+	if dest.Cmp(bigMaxUint64) > 0 {
+		return false
+	}
+	_, ok := dests[dest.Uint64()]
+	return ok
+}
+
+// jitCalculateGasAndSize calculates the required given the opcode and stack items calculates the new memorysize for
+// the operation. This does not reduce gas or resizes the memory.
+func jitCalculateGasAndSize(env Environment, context *Context, caller ContextRef, instr instruction, statedb *state.StateDB, mem *Memory, stack *stack) (*big.Int, *big.Int, error) {
+	var (
+		gas                 = new(big.Int)
+		newMemSize *big.Int = new(big.Int)
+	)
+	err := jitBaseCheck(instr, stack, gas)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// stack Check, memory resize & gas phase
+	switch op := instr.op; op {
+	case SWAP1, SWAP2, SWAP3, SWAP4, SWAP5, SWAP6, SWAP7, SWAP8, SWAP9, SWAP10, SWAP11, SWAP12, SWAP13, SWAP14, SWAP15, SWAP16:
+		n := int(op - SWAP1 + 2)
+		err := stack.require(n)
+		if err != nil {
+			return nil, nil, err
+		}
+		gas.Set(GasFastestStep)
+	case DUP1, DUP2, DUP3, DUP4, DUP5, DUP6, DUP7, DUP8, DUP9, DUP10, DUP11, DUP12, DUP13, DUP14, DUP15, DUP16:
+		n := int(op - DUP1 + 1)
+		err := stack.require(n)
+		if err != nil {
+			return nil, nil, err
+		}
+		gas.Set(GasFastestStep)
+	case LOG0, LOG1, LOG2, LOG3, LOG4:
+		n := int(op - LOG0)
+		err := stack.require(n + 2)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		mSize, mStart := stack.data[stack.len()-2], stack.data[stack.len()-1]
+
+		gas.Add(gas, params.LogGas)
+		gas.Add(gas, new(big.Int).Mul(big.NewInt(int64(n)), params.LogTopicGas))
+		gas.Add(gas, new(big.Int).Mul(mSize, params.LogDataGas))
```

Comment:
I know this was copied from current VM and can be optimised later, but wanted to check: JIT is still sequential right? If so we could re-use pre-allocated big.Int here to avoid allocating these on every instruction?